src/analyses/constant_propagator.cpp

```diff
@@ -181,30 +115,92 @@ void constant_propagator_domaint::transform(
   }
   else if(from->is_function_call())
   {
-    const exprt &function=to_code_function_call(from->code).function();
+    const code_function_callt &function_call=to_code_function_call(from->code);
+    const exprt &function=function_call.function();
+
+    locationt next=from;
+    next++;
 
     if(function.id()==ID_symbol)
     {
-      const irep_idt &identifier=to_symbol_expr(function).get_identifier();
-
-      if(identifier=="__CPROVER_set_must" ||
-         identifier=="__CPROVER_get_must" ||
-         identifier=="__CPROVER_set_may" ||
-         identifier=="__CPROVER_get_may" ||
-         identifier=="__CPROVER_cleanup" ||
-         identifier=="__CPROVER_clear_may" ||
-         identifier=="__CPROVER_clear_must")
+      // called function identifier
+      const symbol_exprt &symbol_expr=to_symbol_expr(function);
+      const irep_idt id=symbol_expr.get_identifier();
+
+      if(to==next)
       {
+        if(id==CPROVER_PREFIX "set_must" ||
+           id==CPROVER_PREFIX "get_must" ||
+           id==CPROVER_PREFIX "set_may" ||
+           id==CPROVER_PREFIX "get_may" ||
+           id==CPROVER_PREFIX "cleanup" ||
+           id==CPROVER_PREFIX "clear_may" ||
+           id==CPROVER_PREFIX "clear_must")
+        {
+          // no effect on constants
+        }
+        else
+        {
+          if(have_dirty)
+            values.set_dirty_to_top(cp->dirty, ns);
+          else
+            values.set_to_top();
+        }
       }
       else
-        values.set_to_top();
+      {
+        // we have an actual call
+
+        // parameters of called function
+        const symbolt &symbol=ns.lookup(id);
+        const code_typet &code_type=to_code_type(symbol.type);
+        const code_typet::parameterst &parameters=code_type.parameters();
+
+        const code_function_callt::argumentst &arguments=
+          function_call.arguments();
+
+        code_typet::parameterst::const_iterator p_it=parameters.begin();
+        for(const auto &arg : arguments)
+        {
+          if(p_it==parameters.end())
+            break;
+
+          const symbol_exprt parameter_expr(p_it->get_identifier(), arg.type());
+          assign_rec(values, parameter_expr, arg, ns);
+
+          ++p_it;
+        }
+      }
     }
     else
-      values.set_to_top();
+    {
+      // unresolved call
+      DATA_INVARIANT(to==next, "Unknown instructions do not branch");
```

Comment:
What is an "Unknown instruction"? If I managed to align braces correctly, isn't this the case of a function call via function pointers?

Reply:
Good catch; I got the bracketing wrong.  There is an implicit assumption ( https://github.com/diffblue/cbmc/issues/751 ) that function pointers will be removed before this is run.  We have an issue to remove this restriction but it won't get done for a little while yet.  I've rephrased this to:

INVARIANT(to==next, "Unresolved call can only be approximated if a skip");